doc/start: Update hardware recommendations.

markhpc opened the pull request:
New hardware recommendations more in-line with current behavior and default values.

Signed-off-by: Mark Nelson <mnelson@redhat.com>

## Checklist
- [ ] References tracker ticket
- [X] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

tchaikov commented on doc/start/hardware-recommendations.rst:
```diff
@@ -266,34 +296,46 @@ and development clusters can run successfully with modest hardware.
 +==============+================+=========================================+
-| ``ceph-osd`` | Processor      | - 1x 64-bit AMD-64                      |
-|              |                | - 1x 32-bit ARM dual-core or better     |
+| ``ceph-osd`` | Processor      | - 1 core minimum                        |
```
we don't test 32bit CPUs. shall we be specific on what architectures do we recommend? because we only run qa suites on AMD64. we do run "make check" on aarch64, but some tests fail now and then.

markhpc replied:
I think we definitely want to remove the language suggesting that we support 32 bit ARM processors. ;)  I'm less sure if we should explicitly state we support AMD64 as it's pretty ubiquitous these days.  I'm a little worried that the more specific we are the more we need to keep these docs up to date and clearly we haven't been very good at it in the past. ;)

tchaikov replied:
i have been contacted recently asking if we "supported" aarch64 . so i am wondering if it'd be more convenience if i can just point them to the document.

zdover23 replied:
@markhpc: If you get an email to me with accurate information, I guarantee that I'll add it to a pull request within 12 hours of receiving the email.

liewegas replied:
i wonder if the 'supported' question si better pointed at a doc about what hardware we build and test for

xahmad commented on doc/start/hardware-recommendations.rst:
```diff
@@ -266,34 +296,46 @@ and development clusters can run successfully with modest hardware.
+|              | RAM            | - 4GB+ per daemon (more is better)      |
+|              |                | - 2-4GB often functions (may be slow)   |
+|              |                | - Less than 2GB not recommended         |
 |              +----------------+-----------------------------------------+
```
Is there an upper limit where adding more RAM does not help anymore?

markhpc replied:
Hi Xahmad,

Really the upper bound is only limited by how many onodes, OMAP, and active data buffers you have.  IE having 2TB of memory for an OSD with a 1TB NVMe drive would be over that upper limit because you'd have more RAM than data over all.  Having 500GB of memory would be more than you'd need to cache onodes (and likely OMAP) but potentially less than you'd need to fully cache a large active data set.  4GB of memory may not be enough to cache active onodes and OMAP depending on various factors (and is much less likely to fully cache the entire active data set).  So generally more memory is better, but the practical benefit typically tapers off fairly quickly after 8-16GB per OSD.

fiskn commented:
Hi Mark,

Just seen this PR and though I would comment on the CPU section. From the testing I did a few years back I came up with the following graph for the amount of CPU required for a 3x replica pool. The overall results more or less line up with whats in the current PR, ie graph below shows 1 IO requires about 1Mhz and PR says 1 core (assume 3Ghz) equals 3000 iops. But I think its useful to be able to make the connection between the clock speed of the core and performance. Especially as the clock speed will ultimately have a large impact on the latency of each IO and delays due to PG contention.

![image](https://user-images.githubusercontent.com/11720958/72246938-c52f0080-35eb-11ea-90e2-97d75276a5c3.png)

Maybe recommending that a high clock speed should be preferred in almost all cases, unless a specific workload is well understood. General strategy would be to: 

- Work out total IO load 
- Convert that to total required Ghz across cluster
- Find CPU where (Ghz*core's) > required total Ghz, preferring higher Ghz per core where possible

Might also be worth hinting that C-State and T-State throttling should be disabled to minimize latency of each IO request?

markhpc commented:
@fiskn It's a little tough because if you have few OSDs on a node you can shoot for high clock speed, but with lots of OSDs you might be better off with more cores.  IE we've had users with 24-32 NVMe drives per node where we believe having lots of slower cores is likely a better solution than fewer faster cores.  On the other hand, if you have say 4-8 NVMe drives in a node, you might be better off with fewer faster cores.

liewegas commented on doc/start/hardware-recommendations.rst:
```diff
@@ -59,8 +59,41 @@ configured to consume.  We recommend 1 GB as a minimum for most systems.  See
+Memory
+======
+
+Bluestore uses it's own memory to cache data rather than relying on the
```
its

liewegas commented on doc/start/hardware-recommendations.rst:
```diff
@@ -59,8 +59,41 @@ configured to consume.  We recommend 1 GB as a minimum for most systems.  See
+the OSD attempts to consume with the ``osd_memory_target`` configuration
+option.  The OSD can only control how much memory is mapped rather than how
+much RSS memory it actually uses.  The kernel is not guaranteed to reclaim
+unmapped memory within any given time frame.  When Transparent Huge Pages are
```
I would move this caveat discussion about THP to after the basics ("set the osd_memory_target option").  And also didn't we add an option (on by default) that disables this automatically?

liewegas commented on doc/start/hardware-recommendations.rst:
```diff
@@ -266,34 +296,46 @@ and development clusters can run successfully with modest hardware.
 |              | Disk Space     |  10 GB per daemon                       |
 |              +----------------+-----------------------------------------+
-|              | Network        |  2x 1GB Ethernet NICs                   |
+|              | Network        |  2x 1GbE+ NICs                          |
```
1x?

liewegas commented on doc/start/hardware-recommendations.rst:
```diff
@@ -266,34 +296,46 @@ and development clusters can run successfully with modest hardware.
 |              | Disk Space     |  1 MB per daemon                        |
 |              +----------------+-----------------------------------------+
-|              | Network        |  2x 1GB Ethernet NICs                   |
+|              | Network        |  2x 1GbE+ NICs                          |
```
1x?

fiskn commented:
@markhpc I think the advice is still the same, for larger NVME nodes faced with the choice between 2 CPU types like Xeon 6252 (24x2.1Ghz=50.4Ghz) or 6254 (18x3.1Ghz=55.8Ghz), both have similar total Ghz and price, but the Xeon 6254 will zip through the serial parts of the code perhaps 50% faster reducing latency and that's without even taking into account its impressive 3.9Ghz all core turbo, if I'm reading the specs right.

markhpc commented:
@fiskn 

In cases like that where you can get two processors with similar IPC and total GHz at rough price parity you are probably correct.   Even in the given example there are tradeoffs though.  The 6252 gives you more L2 (33%) and L3 (44%) cache.  It also theoretically uses less power (200w vs 150w).  That might affect the likelihood of staying in all-core turbo in dual socket machines depending on the cooling setup.  If you are willing to jump up to the platinum 8260Y you can pretty much get the best of both processors but the cost scales roughly linearly.

Probably the bigger question is under what circumstances it makes sense to go with a CPU architecture that gives you more cores with lower clocks/IPC vs an architecture with fewer cores and higher clocks/IPC.  Ultimately I think all of this is probably too nuanced and prone to change for a document like this.  Not to mention that specific vendors are probably going to have very strong opinions about what the right answer is. :)

neha-ojha commented:
@markhpc want to get the first cut of this into octopus?

markhpc commented:
@neha-ojha No one ever approved the PR. :)

liewegas commented:
retest this please

liewegas commented:
@markhpc can you address the 1x NIC comments and move (or remove) the THP discussion? and rebase?

markhpc commented:
@liewegas Ok, I changed it to 1X and ripped out the associated paragraph with that recommendation.  I softened the THP language a bit and moved it below the memory target tuning suggestions.  Also rebased!

liewegas commented:
jenkins render docs

liewegas commented:
```
/home/jenkins-build/build/workspace/ceph-pr-docs/doc/start/hardware-recommendations.rst:71: WARNING: Bullet list ends without a blank line; unexpected unindent.
/home/jenkins-build/build/workspace/ceph-pr-docs/doc/start/hardware-recommendations.rst:74: WARNING: Bullet list ends without a blank line; unexpected unindent.
/home/jenkins-build/build/workspace/ceph-pr-docs/doc/start/hardware-recommendations.rst:78: WARNING: Bullet list ends without a blank line; unexpected unindent.
/home/jenkins-build/build/workspace/ceph-pr-docs/doc/start/hardware-recommendations.rst:81: WARNING: Bullet list ends without a blank line; unexpected unindent.
```
